src/main/java/com/hubspot/imap/client/ImapCodec.java

```diff
@@ -38,11 +38,18 @@ public ImapCodec(ImapClientState clientState) {
   @Override
   protected void encode(ChannelHandlerContext ctx, BaseImapCommand msg, List<Object> out) throws Exception {
     String data = msg.commandString();
-    String tag = clientState.getNextTag();
 
-    trace(ctx, tag, data);
+    if (msg.getRequiresTag()) {
+      String tag = clientState.getNextTag();
```

Comment:
might reduce code duplication to simply assign a `tag` var and use the same send code. i.e:
```java
String tag  = msg.isTagged() ? clientState.getNextTag() : "";
```

Appending the empty string should be fine.